### PR TITLE
create view to add battle

### PR DESCRIPTION
**Pull request by DeborahAlmeida**
## Objective and Motivation
- Create a battle using class-based views
- Create a team
- Create pokemon's team

### CODE REVIEW

File: backend/battle/forms.py [hunk 1]
```diff
@@ -1,14 +1,47 @@
...
+from django import forms
+
+from .models import Battle, PokemonTeam, Team
+from pokemon.models import Pokemon
+
+class TeamForm(forms.ModelForm):
+    class Meta:
+        model = PokemonTeam
+        fields = [
+        'pokemon_1',
```
Comment (rsarai) [inline backend/battle/forms.py]: Identation issue here :see_no_evil:

File: backend/battle/forms.py [hunk 2]
```diff
@@ -1,14 +1,47 @@
...
+        'pokemon_2',
+        'pokemon_3'
+        ]
+
+    pokemon_1 = forms.IntegerField(
```
Comment (rsarai) [inline backend/battle/forms.py]: Please use `ModelMultipleChoiceField` or  `ModelChoiceField`: https://docs.djangoproject.com/en/3.2/topics/forms/modelforms/#field-types

File: backend/battle/forms.py [hunk 3]
```diff
@@ -1,14 +1,47 @@
...
+    #     PokemonTeam.objects.create(team=Team.objects.get(pk=47), pokemon=Pokemon.objects.get(pokemon_id=18), order=1)
+    #     PokemonTeam.objects.create(team=Team.objects.get(pk=47), pokemon=Pokemon.objects.get(pokemon_id=16), order=2)
+    #     PokemonTeam.objects.create(team=Team.objects.get(pk=47), pokemon=Pokemon.objects.get(pokemon_id=15), order=3)
+    #     instance = super().save(commit=False)
+    #     instance.some_flag = True
+    #     if commit:
+    #         instance.save()
+    #         self.save_m2m()
+
+    #     return instance
```
Comment (rsarai) [inline backend/battle/forms.py]: why the comments?

File: backend/battle/urls.py [hunk 4]
```diff
@@ -1,18 +1,15 @@
...
 urlpatterns = [
     path("", Home.as_view(), name="home"),
-    # path('round/new/', RoundNewCreator.as_view(), name='round_new'),
-    # path('invite/', Invite.as_view(), name='invite'),
-    # path('opponent/', Opponent.as_view(), name='opponent'),
-    # path('player2/round', RoundNewOponnent.as_view() , name='round_new2'),
-
+    path("battle/", BattleView.as_view(), name="battle"),
+    path('invite/', Invite.as_view(), name='invite'),
+    path("team/<int:pk>/create", TeamView.as_view(), name="team_create"),
```
Comment (rsarai) [inline backend/battle/urls.py]: The pk here is related to the team or the battle?

File: backend/battle/views.py [hunk 5]
```diff
@@ -9,12 +13,51 @@
...
 
-    def get(self, request):
-        return render(request, self.template_name)
 
+class Invite(TemplateView):
+    template_name = 'battle/invite.html'
+
+
+class BattleView(CreateView):
+    model = Battle
```
Comment (rsarai) [inline backend/battle/views.py:21]: which form is being used here?

File: backend/battle/views.py [hunk 6]
```diff
@@ -9,12 +13,51 @@
...
+class TeamView(CreateView):
+    template_name = "battle/pokemon_form.html"
+    form_class = TeamForm
+    success_url = reverse_lazy("home")
+
+    def get(self, request, pk):
+        self.object = None
+        battle_object = Battle.objects.get(pk=pk)        
+        Team.objects.create(battle=battle_object, trainer=battle_object.creator)
+        Team.objects.create(battle=battle_object, trainer=battle_object.opponent)
```
Comment (rsarai) [inline backend/battle/views.py]: So every time a user access the page new teams will be created? Is this the correct behavior?

File: backend/battle/views.py [hunk 7]
```diff
@@ -9,12 +13,51 @@
...
+        return super().get(request, pk)
+
+
+    # def form_invalid(self, form):
```
Comment (rsarai) [inline backend/battle/views.py]: Please let's try to avoid adding new code with comments

File: backend/battle/views.py [hunk 8]
```diff
@@ -9,12 +13,51 @@
...
+    model = Battle
+    fields = ['opponent']
+
+    def form_valid(self, form):
+        form.instance.creator = self.request.user
+        form.save()
+        return HttpResponseRedirect(reverse_lazy("team_create", args=(form.instance.id,)))
+
+class TeamView(CreateView):
+    template_name = "battle/pokemon_form.html"
```
Comment (rsarai) [inline backend/battle/views.py:37]: To which model is this view related?

**Review by rsarai — CHANGES_REQUESTED** [review]

File: backend/battle/views.py (hunk 8 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py:37]: To Team model, I will fix this

File: backend/battle/views.py (hunk 7 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py]: Sorry about it, I'm gonna delete what is not necessary

File: backend/battle/views.py (hunk 6 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py]: That's wrong, I will think about a new solution. Thanks :)

File: backend/battle/views.py (hunk 5 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py:21]: I didn't make a form for this view, because I thought the CreateView creates a register without a form. I will create a form for this.

File: backend/battle/urls.py (hunk 4 above)
Comment (DeborahAlmeida) [reply backend/battle/urls.py]: It's related to the battle

File: backend/battle/forms.py (hunk 3 above)
Comment (DeborahAlmeida) [reply backend/battle/forms.py]: I was afraid to lose what I wrote, but this is totally wrong. I'm gonna delete it.

File: backend/battle/forms.py (hunk 2 above)
Comment (DeborahAlmeida) [reply backend/battle/forms.py]: Sure, thanks!

File: backend/battle/forms.py (hunk 1 above)
Comment (DeborahAlmeida) [reply backend/battle/forms.py]: Ops, thanks! I'm gonna fix it :D

File: backend/battle/admin.py [hunk 9]
```diff
@@ -1,6 +1,16 @@
...
+    extra = 1
+
+
+class TeamAdmin(admin.ModelAdmin):
+    inlines = (PokemonsInline,)
+
+
 admin.site.register(Battle)
-admin.site.register(Team)
+admin.site.register(Team, TeamAdmin)
```
Comment (rsarai) [inline backend/battle/admin.py:15]: Nice!

File: backend/battle/forms.py [hunk 10]
```diff
@@ -1,24 +1,57 @@
...
+        queryset=Pokemon.objects.all(),
+        required=True,
+    )
+    pokemon_3 = forms.ModelChoiceField(
+        label="Pokemon 3",
+        queryset=Pokemon.objects.all(),
+        required=True,
+    )
+
+    def __init__(self, *args, **kwargs):
```
Comment (rsarai) [inline backend/battle/forms.py:52]: is this required?

File: backend/battle/forms.py [hunk 11]
```diff
@@ -1,24 +1,57 @@
...
+        super().__init__(*args, **kwargs)
+
+    def clean(self):
```
Comment (rsarai) [inline backend/battle/forms.py:57]: A good validation here should be to check if the trainer is or the opponent or the creator of the battle

File: backend/battle/urls.py [hunk 12]
```diff
@@ -1,18 +1,20 @@
...
 
 urlpatterns = [
     path("", Home.as_view(), name="home"),
-    # path('round/new/', RoundNewCreator.as_view(), name='round_new'),
-    # path('invite/', Invite.as_view(), name='invite'),
-    # path('opponent/', Opponent.as_view(), name='opponent'),
-    # path('player2/round', RoundNewOponnent.as_view() , name='round_new2'),
+    path("battle/", BattleView.as_view(), name="battle"),
+    path('invite/', Invite.as_view(), name='invite'),
+    path('battle/list', BattleList.as_view(), name='battle_list'),
```
Comment (rsarai) [inline backend/battle/urls.py]: URLs must finish with `/`

File: backend/battle/views.py [hunk 13]
```diff
@@ -1,98 +1,57 @@
...
-# from .battles.base_stats import get_total_point_pokemon
-
-
-class Home(View):
+from django.http import HttpResponseRedirect
+from django.db.models import Q
+from django.views.generic import TemplateView, CreateView, ListView, DetailView
+from .models import Battle, Team
+from django.urls import reverse_lazy
+from .forms import TeamForm, BattleForm
```
Comment (rsarai) [inline backend/battle/views.py]: I try to avoid relative imports, it can cause problems when copying and pasting code or when changing files location

File: backend/battle/urls.py [hunk 14]
```diff
@@ -1,18 +1,20 @@
...
+    path("team/<int:pk>/create/<int:user>", TeamView.as_view(), name="team_create"),
```
Comment (rsarai) [inline backend/battle/urls.py]: You don't need to pass the user here. You can use the logged user inside the view

File: backend/battle/views.py [hunk 15]
```diff
@@ -1,98 +1,57 @@
...
+
+class BattleView(CreateView):
+    model = Battle
+    template_name = 'battle/battle_form.html'
+    form_class = BattleForm
+
+    def form_valid(self, form):
+        form.instance.creator = self.request.user
+        form.save()
+        return HttpResponseRedirect(reverse_lazy("team_create", args=(form.instance.id, 1, )))
```
Comment (rsarai) [inline backend/battle/views.py]: I suggested to change the url params by removing the user's id

File: backend/battle/views.py [hunk 16]
```diff
@@ -1,98 +1,57 @@
...
+
+
+class TeamView(CreateView):
+    model = Team
+    template_name = "battle/pokemon_form.html"
+    form_class = TeamForm
+    success_url = reverse_lazy("invite")
+
+    def get_initial(self):
+        super(TeamView, self).get_initial()
```
Comment (rsarai) [inline backend/battle/views.py]: since `get_initial` returns an dict you should do: `self.initial = super(TeamView, self).get_initial()`

File: backend/battle/views.py [hunk 17]
```diff
@@ -1,98 +1,57 @@
...
+        self.initial = {"battle": self.kwargs['pk'], "user": self.kwargs['user']}
```
Comment (rsarai) [inline backend/battle/views.py]: The initial fields should be the same ones as the form. In this case, `trainer` and `battle`.

Comment (rsarai) [reply backend/battle/views.py]: Also, instead of using the `self.kwargs['pk']` you can pass the battle object.

File: backend/battle/forms.py [hunk 18]
```diff
@@ -1,24 +1,57 @@
...
+        cleaned_data = super().clean()
+        return cleaned_data
+
+    def save(self, commit=True):
+        battle_object = Battle.objects.get(pk=self.initial['battle'])
+        if self.initial['user'] == 1:
```
Comment (rsarai) [inline backend/battle/forms.py]: At this point, the user passed to the form should be the correct user. The validations on `clean()` should ensure that

File: backend/battle/forms.py [hunk 19]
```diff
@@ -1,24 +1,57 @@
...
+            team = Team.objects.create(battle=battle_object, trainer=battle_object.creator)
+        else:
+            team = Team.objects.create(battle=battle_object, trainer=battle_object.opponent)
+        PokemonTeam.objects.create(team=team, pokemon=Pokemon.objects.get(pokemon_id=18), order=1)
+        PokemonTeam.objects.create(team=team, pokemon=Pokemon.objects.get(pokemon_id=16), order=2)
+        PokemonTeam.objects.create(team=team, pokemon=Pokemon.objects.get(pokemon_id=15), order=3)
```
Comment (rsarai) [inline backend/battle/forms.py]: You should use the data selected from the `pokemon_1`, `pokemon_2`, and `pokemon_3` fields.

**Review by rsarai — CHANGES_REQUESTED** [review]

File: backend/battle/forms.py (hunk 10 above)
Comment (DeborahAlmeida) [reply backend/battle/forms.py:52]: It's not, I think I overwrote without use

File: backend/battle/forms.py (hunk 11 above)
Comment (DeborahAlmeida) [reply backend/battle/forms.py:57]: I do this validation on def save() but I think it's not the right place to make this. I'm gonna make on def clean()

File: backend/battle/urls.py (hunk 12 above)
Comment (DeborahAlmeida) [reply backend/battle/urls.py]: thanks!

File: backend/battle/views.py (hunk 13 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py]: thank you!

File: backend/battle/views.py (hunk 15 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py]: Okay :)

File: backend/battle/views.py (hunk 16 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py]: hmm I see

File: backend/battle/views.py (hunk 17 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py]: thanks :D

File: backend/battle/battles/battle.py [hunk 20]
```diff
@@ -1,29 +1,47 @@
...
+def battleRunning(team):
+    battle_info = team.battle
+    pokemon_team_opponent = PokemonTeam.objects.filter(team=team)
+    team_creator = Team.objects.get(battle=battle_info, trainer=battle_info.creator)
+    pokemon_team_creator = PokemonTeam.objects.filter(team=team_creator)
+    winner = get_points(pokemon_team_opponent, pokemon_team_creator)
+    return winner
+
+
+def get_points(pokemons_creator, pokemons_opponent):
```
Comment (rsarai) [inline backend/battle/battles/battle.py]: I think the name here is misleading, you are not getting the points, you are fetching the winner.

File: backend/battle/battles/battle.py [hunk 21]
```diff
@@ -1,29 +1,47 @@
...
-# def message_error():
-#     message = "ERROR: PKNs you selected sum more than 600 points, please choose again"
-#     return message
+from battle.models import Battle
+from battle.battles.rounds import roundRunning
+from battle.models import PokemonTeam, Team
+from battle.battles.email import result_battle
+
+
+def battleRunning(team):
```
Comment (rsarai) [inline backend/battle/battles/battle.py]: method names should be `snake_case`

File: backend/battle/battles/battle.py [hunk 22]
```diff
@@ -1,29 +1,47 @@
...
+    pokemon_team_creator = PokemonTeam.objects.filter(team=team_creator)
+    winner = get_points(pokemon_team_opponent, pokemon_team_creator)
+    return winner
+
+
+def get_points(pokemons_creator, pokemons_opponent):
+    battle_rounds = [
+        [pokemons_creator[0].pokemon, pokemons_opponent[0].pokemon],
+        [pokemons_creator[1].pokemon, pokemons_opponent[1].pokemon],
+    ]
```
Comment (rsarai) [inline backend/battle/battles/battle.py]: I'm confused, shouldn't this be three pokemon? Also, you can pass the teams and use `team_creator.pokemons` and `team_opponent.pokemons`

File: backend/battle/battles/battle.py [hunk 23]
```diff
@@ -1,29 +1,47 @@
...
+
+def validate_sum_pokemons(pokemons):
+    total_points = 0
+    for pokemon in pokemons:
+        pokemon_point = pokemon.attack + pokemon.defense + pokemon.hp
+        total_points += pokemon_point
+    return total_points <= 600
+
+
+def setWinner(winner, battle):
```
Comment (rsarai) [inline backend/battle/battles/battle.py]: method names should be `snake_case`

File: backend/battle/forms.py [hunk 24]
```diff
@@ -1,24 +1,76 @@
...
 
+class BattleForm(forms.ModelForm):
+    class Meta:
+        model = Battle
+        fields = ['opponent', ]
 
-# class BattleForm(forms.ModelForm):
+    def clean(self):
+        cleaned_data = super().clean()
+        if cleaned_data['opponent'] == self.initial['creator']:
```
Comment (rsarai) [inline backend/battle/forms.py]: Why not do the following?
```python
if cleaned_data['opponent'] == self.cleaned_data['creator']:
```

File: backend/battle/forms.py [hunk 25]
```diff
@@ -1,24 +1,76 @@
...
+                cleaned_data['pokemon_1'],
+                cleaned_data['pokemon_2'],
+                cleaned_data['pokemon_3']
+            ]
+        )
+        obj_battle = cleaned_data['battle']
 
-# class RoundForm2(forms.ModelForm):
+        if cleaned_data['trainer'] != obj_battle.creator:
+            if cleaned_data['trainer'] != obj_battle.opponent:
```
Comment (rsarai) [inline backend/battle/forms.py]: This could be replaced by:

```python
if cleaned_data['trainer'] != obj_battle.creator and cleaned_data['trainer'] != obj_battle.opponent:

```

File: backend/battle/templates/battle/battle_list.html [hunk 26]
```diff
@@ -0,0 +1,109 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <title>Pokebattle</title>
+  <style>
+    body{
+        padding: 0px;
+        background: #1488CC;  /* fallback for old browsers */
+        font-family: Arial, Helvetica, sans-serif;
+
```
Comment (rsarai) [inline backend/battle/templates/battle/battle_list.html]: extra spaces here

File: backend/battle/views.py [hunk 27]
```diff
@@ -1,98 +1,73 @@
...
+    template_name = 'battle/invite.html'
+
+
+class BattleView(CreateView):
+    model = Battle
+    template_name = 'battle/battle_form.html'
+    form_class = BattleForm
+
+    def get_initial(self):
+        obj_creator = get_object_or_404(User, email=self.request.user)
```
Comment (rsarai) [inline backend/battle/views.py]: This is not required, the logged user is already saved on the request, you can just do: `self.initial = {"creator": self.request.user}`

File: backend/battle/views.py [hunk 28]
```diff
@@ -1,98 +1,73 @@
...
+        self.initial = {"creator": obj_creator}
+        return self.initial
+
+    def form_valid(self, form):
+        form.instance.creator = self.request.user
```
Comment (rsarai) [inline backend/battle/views.py]: is this required? Since you are already passing on the `get_initial`

File: backend/battle/views.py [hunk 29]
```diff
@@ -1,98 +1,73 @@
...
+
+class TeamView(CreateView):
+    model = Team
+    template_name = "battle/pokemon_form.html"
+    form_class = TeamForm
+    success_url = reverse_lazy("invite")
+
+    def get_initial(self):
+        obj_battle = get_object_or_404(Battle, pk=self.kwargs['pk'])
+        obj_trainer = get_object_or_404(User, email=self.request.user)
```
Comment (rsarai) [inline backend/battle/views.py]: Same as above, you don't need to fetch the user since it's already loaded on the request

File: backend/battle/views.py [hunk 30]
```diff
@@ -1,98 +1,73 @@
...
+
+    def form_valid(self, form):
+        team = form.save()
+        if team.battle.creator == team.trainer:
+            return HttpResponseRedirect(reverse_lazy("invite"))
+        if team.battle.opponent == team.trainer:
+            result = battleRunning(team)
+            setWinner(result, team.battle)
+            return HttpResponseRedirect(reverse_lazy("home"))
+        return True
```
Comment (rsarai) [inline backend/battle/views.py]: This redirect logic is a bit misleading. What happens when the opponent creates a team before the creator? I think you should only run the battle if both teams are created.

**Review by rsarai — CHANGES_REQUESTED** [review]

File: backend/battle/forms.py (hunk 24 above)
Comment (DeborahAlmeida) [reply backend/battle/forms.py]: It's because I don't have the creator value on cleaned_data by default. I put the creator value on form_valid(). So to solve this I pass the self.request.user by get_initial()

File: backend/battle/views.py (hunk 28 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py]: Nop :)

File: backend/battle/views.py (hunk 29 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py]: But when I send the self.request.user, it gives me an error because it's not the object, how can I solve this?

File: backend/battle/forms.py (hunk 24 above)
Comment (DeborahAlmeida) [reply backend/battle/forms.py]: I fixed it :) thanks

File: backend/battle/forms.py (hunk 25 above)
Comment (DeborahAlmeida) [reply backend/battle/forms.py]: Oh great! Thank you :)

File: backend/battle/battles/battle.py [hunk 31]
```diff
@@ -1,29 +1,42 @@
...
-# def message_error():
-#     message = "ERROR: PKNs you selected sum more than 600 points, please choose again"
-#     return message
+from battle.battles.rounds import run_round
+from battle.battles.email import result_battle
+
+
+def run_battle(battle):
+    teams = battle.teams.all()
+    assert teams.count() == 2, "Unexpected error. Battle has more than two teams."
```
Comment (fjsj) [inline backend/battle/battles/battle.py]: Just a small optimization you may want to learn here.
If you do `battle.teams.all()` then `(teams[0], teams[1])` you do 1 DB query. But `teams.count()` does another DB query.

Since you're going to fetch anyway the team items (0 and 1), you can always ensure 1 DB query only by doing this:

```
teams = list(battle.teams.all())
assert len(teams) == 2, "Unexpected error. Battle has more than two teams."
team_winner = get_winner_for(teams[0], teams[1])
```

File: backend/battle/battles/email.py [hunk 32]
```diff
@@ -1,20 +1,13 @@
...
     send_templated_mail(
         template_name='results_battle',
         from_email='deborahmendonca6@gmail.com',
-        recipient_list=['deborah.mendonca@vinta.com.br'],
+        recipient_list=[battle.creator, battle.opponent],
         context={
-            'winner': winner,
+            'winner': battle.winner,
         },
         # Optional:
```
Comment (fjsj) [inline backend/battle/battles/email.py]: Please remove those comments

File: backend/battle/models.py [hunk 33]
```diff
@@ -24,7 +24,7 @@ class Team(models.Model):
 class PokemonTeam(models.Model):
     team = models.ForeignKey(Team, on_delete=models.CASCADE, related_name="teams")
     pokemon = models.ForeignKey(Pokemon, on_delete=models.CASCADE, related_name="+")
-    order = models.PositiveIntegerField()
+    order = models.PositiveIntegerField(null=True)
```
Comment (fjsj) [inline backend/battle/models.py]: Sorry if you already discussed this with Rebeca, but does this really need to be `null=True`?

File: backend/battle/views.py [hunk 34]
```diff
@@ -1,98 +1,71 @@
...
+    template_name = 'battle/invite.html'
+
+
+class BattleView(CreateView):
+    model = Battle
+    template_name = 'battle/battle_form.html'
+    form_class = BattleForm
+
+    def get_initial(self):
+        obj_creator = get_object_or_404(User, email=self.request.user)
```
Comment (fjsj) [inline backend/battle/views.py]: I think this is still wrong, no? `obj_creator = self.request.user` doesn't work?

File: backend/pokemon/models.py [hunk 35]
```diff
@@ -5,6 +5,6 @@
 class Pokemon(models.Model):
     pokemon_id = models.IntegerField(null=True)
     name = models.CharField(max_length=200, null=True)
-    attack = models.CharField(max_length=200, null=True)
-    defense = models.CharField(max_length=200, null=True)
-    hp = models.CharField(max_length=200, null=True)
+    attack = models.IntegerField(null=True)
+    defense = models.IntegerField(null=True)
+    hp = models.IntegerField(null=True)
```
Comment (fjsj) [inline backend/pokemon/models.py]: Sorry if you already discussed this with Rebeca, but does all fields here need to be `null=True`?

**Review by fjsj — CHANGES_REQUESTED** [review]
Please check comments

File: backend/battle/battles/battle.py (hunk 31 above)
Comment (DeborahAlmeida) [reply backend/battle/battles/battle.py]: Great! I didn't know about it. I'm going to improve this :) Thank you

File: backend/battle/battles/email.py (hunk 32 above)
Comment (DeborahAlmeida) [reply backend/battle/battles/email.py]: Okay, thank you for remember me :)

File: backend/battle/models.py (hunk 33 above)
Comment (DeborahAlmeida) [reply backend/battle/models.py]: Previously it was a requirement because I used to create a register on PokemonTeam without order and then I updated it. But now it's not like this, so I'm going to change it. Thank you one more time :)

File: backend/battle/views.py (hunk 34 above)
Comment (DeborahAlmeida) [reply backend/battle/views.py]: Actually, it works fine, I didn't realize any error with this. Maybe it has something I didn't recognize. Can you explain why this could cause an error?

File: backend/pokemon/models.py (hunk 35 above)
Comment (DeborahAlmeida) [reply backend/pokemon/models.py]: Actually, they don't. I'll change this. Thanks!

File: backend/battle/views.py (hunk 34 above)
Comment (fjsj) [reply backend/battle/views.py]: maybe you did get an error when you accessed this page without logging in beforehand.
but either way, it would fail without this as well

Comment (DeborahAlmeida) [reply backend/battle/views.py]: thank you for the explanation!

**Review by rsarai — APPROVED** [review]
